### ISSUE TITLE
[RFE] Localized login dialog on non-English system

### ISSUE THREAD

**Issue by AsciiWolf** [neutrinolabs/xrdp#1637]
It seems that the xrdp login prompt is English only. It would be great if it could be made translatable.

**Comment by matt335672** [comment]
Definitely related to #2059, and possibly the same.

**Comment by AsciiWolf** [comment]
I don't think that this issue is the same. This issue is about i18n support in xrdp dialog(s) while the other one is about user-customized strings.

**Comment by matt335672** [comment]
I agree they're not exactly the same, but I think there's definitely some overlap. Or do you not think there is?

I was thinking that the same mechanism could be used to address both issues. I could very well be wrong though, so I'd appreciate your thoughts on this.

**Comment by AsciiWolf** [comment]
I am not sure, however localization is usually handled differently than custom strings/UI parts.

**Comment by matt335672** [comment]
Yes you're right - many applications deliver a separate file, or files for this.

If the login dialog offered a way to customise the language dynamically as some do, I'd say this is definitely a separate issue. The xrdp dialog doesn't however. So could we not achieve a similar result by using a mechanism like this:-

```
password[Password]=...
```

for capitalisation, or
 
```
password[Heslo]=...
```

for internationalisation?

Please feel free to tell me I'm wrong here. I won't be offended, and I'd like to work out the best way forward.

**Comment by AsciiWolf** [comment]
We could. But that would be a workaround that requires manual steps (which is still better than nothing), not a proper (automatic) solution. :-)

**Comment by matt335672** [comment]
That's a fair point.

One other thing that occurred to me was that we could maybe use info in the [TS_INFO_PACKET](https://docs.microsoft.com/en-us/openspecs/windows_protocols/ms-rdpbcgr/732394f5-e2b5-4ac5-8a0a-35345386b0d1) to work out the string to use which best suits the remote user. If the flags field contains the INFO_UNICODE flag, the CodePage field should contain a language identifier. 

I've just tried it here (in the UK) with some additional logging, and from a Windows PC I am indeed getting '0x0809' which according to [[MS-LCID]](https://docs.microsoft.com/en-us/openspecs/windows_protocols/ms-lcid/70feba9f-294e-491e-b6eb-56532684c37f) is en-GB. I think this comes from the Windows registry value "HKLS\SYSTEM\CurrentControlSet\Control\Nls\Language\InstallLanguage"

So in summary, I now agree with you. This should be a different option from #2059. What we should be doing is mapping the client language identifier (if available) to different strings for the field names. I'm guessing from your profile that you'd be getting a language identifier of 0x0405 which maps to a language tag of cs-CZ.

Does that sound reasonable?

**Comment by AsciiWolf** [comment]
Yep, this does sound good, thanks! :-)

**Comment by andrewdbate** [comment]
Hope you don't mind me commenting.

So is the idea that labels in the logon screen are associated with identifiers, and then these identifiers are mapped to strings in translation files?

For example, currently the `username` key in the configuration file is shown as `username` on the login screen.

Following @matt335672's excellent idea, with the new scheme the language would be detected based on the client. If the language was `en-GB` then the value of `username` would be looked up in the `en-GB` translation file, and the mapping would be `username = "Username:"`. Whereas if the client's language was `de-DE` then the translation file would contain the mapping `username = "Benutzername:"`. (I am assuming that we would need to add the trailing colons (if desired to match the presentation in the Windows Remote Desktop client) to the translation file in case trailing colons did not make sense in some languages?)

One final thought for what it's worth: It could be argued that it might not make sense to set the login screen language based on the client, because the language used in the desktop session will be determined/set elsewhere. However, I would not agree with that, because (1) from a practical point-of-view this is clearly useful, and (2) if we had support for NLA (https://github.com/neutrinolabs/xrdp/issues/256) then the logon prompt would be on the client anyway, and the language used on the prompt would be set by the client.

**Comment by matt335672** [comment]
Not at all - the more comments we get which are roughly in agreement, the more I feel that we're going in the right direction.

The translations can be part of the `/etc/xrdp` directory hierarchy like the keyboard translations. That makes them editable by the user as required, and also makes it easier to for contributors who speak other languages to add translations, and then submit a PR.

**Comment by matt335672** [comment]
I've had a bit more of a look into this, and there's a fair bit to do to get this working well. I've no idea when I (or anyone else) will get round to looking at it, so here are some notes:-
1) The login screen currently uses a fixed size for the label width (`ls_label_width`). This will need to be made variable somehow, so we can cope with larger and smaller label widths automatically.
2) The font is currently rendered from the xrdp file `sans-10.fv1`. This is a bitmap representation of DejaVu Sans 10 point (see [this post](https://github.com/neutrinolabs/xrdp/issues/232#issuecomment-88187147)) with Unicode codepoints up to U+4E00.
3) `sans-10.fv1` can represent many, but not all characters that may be required for internationalisation.
4) `sans10-fv1` is a bitmap, and no anti-aliasing is performed when it is rendered.

The last two points are significant, in that intricate characters in non-European languages are going to look pretty dreadful, even if they can be represented at all.

I suspect the best way to proceed with this is to solve the hard problem first, which is the font rendering. Once this is done, the login screen will look a lot better anyway. The remaining issues are not at all complex.

First step is to identify a font rendering library which is going to be reasonably simple to retro-fit to xrdp, and which has a lifecycle which we can depend on. An obvious candidate is [freetype](https://freetype.org/) which is used by the X11 XFT extension. I've not really looked at any other alternatives though.

The [license](https://dejavu-fonts.github.io/License.html) of Deja Vu Sans is such that we can distribute a copy with xrdp if we are concerned about the availability of the font. On my Ubuntu 20.04 system, this font is around 750K in size which compares reasonably well with `sans-10.fv1` at 650K.

As ever, comments welcome.

**Comment by andrewdbate** [comment]
Given the substantial amount of work that this would require, would it be better to focus on implementing NLA (https://github.com/neutrinolabs/xrdp/issues/256) which is considered more modern and secure, and would also sidestep all these issues? If we had NLA support, then the login screen could be considered legacy. (However, I know this is easy for me to say because I am not the one who would be implementing it.)

**Comment by matt335672** [comment]
It's a good question.

Sadly, NLA is even more work. It can't easily be implemented in the current authentication architecture. We're refactoring the architecture (see #1961) and moving xrdp and xrdp-sesman on to the same machine over time. When we've done that, things like NLA and auto session reconnection will be possible.

**Comment by rivimey** [comment]
I came looking here because I wanted to control the font size on my 4K monitor. Ideally the font name too, but size was my desire. I now see that the font is hardcoded (literally!) and do my hopes of a simple solution are dashed.

Does the mention above of the need for additional glyphs, and antialiasing, allow for this issue to switch to a vector font library such as freetype? That would fix many problems in this area I think.

All additional flexibility in the rendering of the login screen would be lovely - e.g as well as i18n as outlined above (seems good to me), provide for a fixed-language override so arbitrary strings can be used, not just using the language translation method. Also set the height of the text input boxes.

**Comment by matt335672** [comment]
The font size problem is I think a separate issue. All the sizes on the login box are hard-coded, and simply fixing the font size isn't going to solve this one.

You could change all the hard-coded sizes in addition to the font which would be fine for your 4K monitor, but then would look bad if you tried logging in from a different source. Ideally the code would work out the monitor DPI from the information presented by the client and work with that.

**Comment by Mogaba** [comment]
So are there any news on this issue?

**Comment by AsciiWolf** [comment]
@matt335672 Sorry for the ping, but has there been any progress on this? Thanks!

**Comment by matt335672** [comment]
Not so far I'm afraid.
